docs/rebase.md

```diff
@@ -0,0 +1,125 @@
+
+## Rebasing the go.mod file and vendoring
+
+In MicroShift's `go.mod` file, we only explicitly add the `require` directives needed by MicroShift itself (marked with the comment `// microshift`) whereas we let `go mod tidy` figure out the direct and indirect requirements for the embedded components. For the rebase, the focus is therefore on the `replace` directives.
+
+When resolving version mismatches between modules used by different embedded components, the general heuristic is to start from a minimal subset of ReplaceDirectives from `o/k` (only those `go mod tidy` actually tries to find) to ensure these are consistent, then add etcd's, openshift-apiserver's and openshift-controller-manager's dependencies.
+
+The `rebase.sh` script automates updating the modulepaths (e.g. rewriting local paths like `./staging` to the global paths) and versions, but it needs hints which component's version to pick. We encode these hints as keywords in comments in MicroShift's `go.mod` file. Each replacement modulepath and versions are picked as follows:
+
+```
+// from $COMPONENT          picks from the $COMPONENT's go.mod
+// release $COMPONENT       picks from the OpenShift's release image
+// override                 keep the current version / do not replace
+```
+
+Run the following to rebase the `go.mod`:
+
+```
+./scripts/rebase.sh go.mod
+go mod tidy
+git add go.mod go.sum
+git commit -m "update go.mod"
+```
+
+As we replace `k8s.io/apiserver` with the new modulepath and version picked from `openshift-apiserver` (i.e. `github.com/openshift/kubernetes-apiserver`) and this repo only carries a subset of patches that OpenShift carries in the corresponding o/k staging dir (i.e. `github.com/openshift/kubernetes/staging/src/k8s.io/apiserver`), we need to add a few missing patches to the repo after each vendoring. These patches are stored in `scripts/rebase_patches`.
+
+To update the vendoring run:
+
+```
+go mod vendor
+git apply scripts/rebase_patches/*
+script/rebase.sh generated-apis
+git add vendor
+git commit -m "update vendoring"
+```
+
+These patches are produced by identifying the missing patches [TODO: describe process] and transforming those patches to apply against `/vendor` instead of o/k's staging dir [TODO: describe process]. This process requires some manual work but should be necessary rarely (likely after OpenShift rebases onto a upstream version).
+
+## Rebasing `Makefile`, `Dockerfiles`, and `.spec` file
+When updating to a new minor version of OpenShift, update the `Makefile` [here](https://github.com/openshift/microshift/blob/main/Makefile#L16) and [here](https://github.com/openshift/microshift/blob/main/Makefile#L55-L66) and also check whether the Dockerfiles need a new base image (e.g. [this](https://github.com/openshift/microshift/blob/main/packaging/images/openshift-ci/Dockerfile)) and the RPM `.spec` file needs updates (e.g. [this](https://github.com/openshift/microshift/blob/main/packaging/rpm/microshift.spec))
```

Comment:
The lines in Makefile are likely to go out of sync quickly. Let's use a description of those places?

Reply:
Good point. Fixed this.